repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasBusinessMetadataDefStoreV2.java

```diff
@@ -363,44 +369,91 @@ private AtlasBusinessMetadataDef toBusinessMetadataDef(AtlasVertex vertex) throw
+        }
+
+        String qualifiedName      = AtlasStructType.AtlasAttribute.getQualifiedAttributeName(bmDef, attributeDef.getName());
+        String vertexPropertyName = AtlasStructType.AtlasAttribute.generateVertexPropertyName(bmDef, attributeDef, qualifiedName);
+
+        boolean isPresent;
+        long startTime = System.currentTimeMillis();
+
+        if (Boolean.TRUE.equals(attributeDef.getIsIndexable())) {
+            isPresent = isBusinessAttributePresent(vertexPropertyName, applicableTypes);
+        } else {
+            isPresent = isBusinessAttributePresentInGraph(vertexPropertyName, applicableTypes);
+        }
+
+        if (LOG.isDebugEnabled()) {
+            LOG.info("Reference check for attribute {} took {} ms. Found: {}",
+                    attributeDef.getName(), (System.currentTimeMillis() - startTime), isPresent);
+        }
+
+        if (isPresent) {
+            throw new AtlasBaseException(AtlasErrorCode.TYPE_HAS_REFERENCES, bmDef.getName());
         }
     }
 
     private boolean isBusinessAttributePresent(String attrName, Set<String> applicableTypes) throws AtlasBaseException {
         SearchParameters.FilterCriteria criteria = new SearchParameters.FilterCriteria();
-
         criteria.setAttributeName(attrName);
-        criteria.setOperator(SearchParameters.Operator.NOT_EMPTY);
+        criteria.setOperator(SearchParameters.Operator.NOT_NULL);
 
         SearchParameters.FilterCriteria entityFilters = new SearchParameters.FilterCriteria();
-
         entityFilters.setCondition(SearchParameters.FilterCriteria.Condition.OR);
         entityFilters.setCriterion(Collections.singletonList(criteria));
 
         SearchParameters searchParameters = new SearchParameters();
-
         searchParameters.setTypeName(String.join(SearchContext.TYPENAME_DELIMITER, applicableTypes));
         searchParameters.setExcludeDeletedEntities(true);
         searchParameters.setIncludeSubClassifications(false);
         searchParameters.setEntityFilters(entityFilters);
-        searchParameters.setAttributes(Collections.singleton(attrName));
-        searchParameters.setOffset(0);
+        searchParameters.setAttributes(Collections.emptySet());
         searchParameters.setLimit(1);
 
         AtlasSearchResult atlasSearchResult = entityDiscoveryService.searchWithParameters(searchParameters);
+        return atlasSearchResult != null && CollectionUtils.isNotEmpty(atlasSearchResult.getEntities());
+    }
+
+    private boolean isBusinessAttributePresentInGraph(String vertexPropertyName, Set<String> applicableTypes) {
+        if (graph == null || CollectionUtils.isEmpty(applicableTypes)) {
+            return false;
+        }
+
+        try {
+            for (String typeName : applicableTypes) {
 
-        return CollectionUtils.isNotEmpty(atlasSearchResult.getEntities());
+                Iterable<AtlasVertex> vertices = graph.query()
+                        .has(Constants.ENTITY_TYPE_PROPERTY_KEY, typeName)
```

Comment:
Is it necessary to add `ENTITY_TYPE_PROPERTY_KEY` filter here? If yes, note that all subTypes of types in`applicableTypes` should be checked as well; consider a business attribute associated with a type like `DataSet`.